stereo_image_proc/src/stereo_image_proc/point_cloud_node.cpp

```diff
@@ -0,0 +1,268 @@
+//    disclaimer in the documentation and/or other materials provided
+//    with the distribution.
+//  * Neither the name of the copyright holder nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+// COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#include <image_geometry/stereo_camera_model.h>
+#include <image_transport/image_transport.h>
+#include <image_transport/subscriber_filter.h>
+#include <message_filters/subscriber.h>
+#include <message_filters/synchronizer.h>
+#include <message_filters/sync_policies/approximate_time.h>
+#include <message_filters/sync_policies/exact_time.h>
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp_components/register_node_macro.hpp>
+#include <rcutils/logging_macros.h>
+#include <sensor_msgs/image_encodings.hpp>
+#include <sensor_msgs/msg/point_cloud2.hpp>
+#include <sensor_msgs/point_cloud2_iterator.hpp>
+#include <stereo_msgs/msg/disparity_image.hpp>
+
+#include <limits>
+#include <memory>
+#include <string>
+
+namespace stereo_image_proc
+{
+
+class PointCloudNode : public rclcpp::Node
+{
+public:
+  explicit PointCloudNode(const rclcpp::NodeOptions & options);
+
+private:
+  // Subscriptions
+  image_transport::SubscriberFilter sub_l_image_;
+  message_filters::Subscriber<sensor_msgs::msg::CameraInfo> sub_l_info_, sub_r_info_;
+  message_filters::Subscriber<stereo_msgs::msg::DisparityImage> sub_disparity_;
+  using ExactPolicy = message_filters::sync_policies::ExactTime<
+    sensor_msgs::msg::Image,
+    sensor_msgs::msg::CameraInfo,
+    sensor_msgs::msg::CameraInfo,
+    stereo_msgs::msg::DisparityImage>;
+  using ApproximatePolicy = message_filters::sync_policies::ApproximateTime<
+    sensor_msgs::msg::Image,
+    sensor_msgs::msg::CameraInfo,
+    sensor_msgs::msg::CameraInfo,
+    stereo_msgs::msg::DisparityImage>;
+  using ExactSync = message_filters::Synchronizer<ExactPolicy>;
+  using ApproximateSync = message_filters::Synchronizer<ApproximatePolicy>;
+  std::shared_ptr<ExactSync> exact_sync_;
+  std::shared_ptr<ApproximateSync> approximate_sync_;
+
+  // Publications
+  // TODO(jacobperron): Uncomment when we can be notified of subscriber status
```

Comment:
Can't you check with http://docs.ros2.org/dashing/api/rclcpp/classrclcpp_1_1Node.html#a7aa2ed22450d2f3300ea5b879acd73b8?

Reply:
We can poll for subscriptions, but I think we'd prefer this to be event driven.
I wasn't aware of the graph event API, I'll see if we can use it.

Reply:
Ah, well what I think would accomplish the exact same thing would be to have the callback just check if any of the output topics have any subscribers, else just return without doing any work.

The `connectCb()` is nice so you don't have to process any of the callbacks that would just be exiting, but right now, Dashing/Eloquent have an issue with unsubscribing and resubscribing to a topic causing a crash (we have some navigation / rclcpp tickets around), so I'd rather do it in a non-crashing way.

Reply:
Sure, that sounds reasonable. For posterity, could you reference an issue regarding problems with resubscribing? Thanks!

Reply:
https://github.com/ros-planning/navigation2/issues/1363

https://github.com/ros2/rmw_fastrtps/issues/341

Reply:
See 3e40a3d